src/middleware/insights.ts

```diff
@@ -0,0 +1,114 @@
+  const insightsClient =
+    _insightsClient === null ? (noop as InsightsClient) : _insightsClient;
+
+  return ({ instantSearchInstance }) => {
+    insightsClient('_get', '_hasCredentials', (hasCredentials: boolean) => {
+      if (!hasCredentials) {
+        const [appId, apiKey] = getAppIdAndApiKey(instantSearchInstance.client);
+        insightsClient('_get', '_userToken', (userToken: string) => {
+          warning(
+            !userToken,
+            `You set userToken before \`createInsightsMiddleware()\` and it is ignored.
+Please set the token after the \`createInsightsMiddleware()\` call.
+
+createInsightsMiddleware({ /* ... */ });
+
+insightsClient('setUserToken', 'your-user-token');
+// or
+aa('setUserToken', 'your-user-token');
+            `
+          );
+        });
+        insightsClient('init', { appId, apiKey });
+      }
+    });
+
+    return {
+      onStateChange() {},
+      subscribe() {
+        const setUserTokenToSearch = (userToken?: string) => {
+          // At the time this middleware is subscribed, `mainIndex.init()` is already called.
+          // It means `mainIndex.getHelper()` exists.
+          if (userToken) {
+            instantSearchInstance.mainIndex
+              .getHelper()!
+              .setQueryParameter('userToken', userToken);
+          }
+        };
+
+        instantSearchInstance.mainIndex
+          .getHelper()!
+          .setQueryParameter('clickAnalytics', true);
+
+        if (hasInsightsClient) {
+          // When `aa('init', { ... })` is called, it creates an anonymous user token in cookie.
+          // We can set it as userToken.
+          setUserTokenToSearch(getInsightsAnonymousUserToken());
+        }
+
+        if (Array.isArray((insightsClient as any).queue)) {
+          // Context: The umd build of search-insights is asynchronously loaded by the snippet.
+          //
+          // When user calls `aa('setUserToken', 'my-user-token')` before `search-insights` is loaded,
+          // it is stored in `aa.queue` and we are reading it to set userToken to search call.
+          // This queue is meant to be consumed whenever `search-insights` is loaded and when it runs `processQueue()`.
+          // But the reason why we handle it here is to prevent the first search API from being triggered
+          // without userToken because search-insights is not loaded yet.
+          (insightsClient as any).queue.forEach(([method, firstArgument]) => {
+            if (method === 'setUserToken') {
+              setUserTokenToSearch(firstArgument);
+            }
+          });
+        }
+
+        // This updates userToken which is set explicitly by `aa('setUserToken', userToken)`
+        insightsClient('onUserTokenChange', setUserTokenToSearch, {
+          immediate: true,
+        });
+      },
+      unsubscribe() {
+        insightsClient('onUserTokenChange', undefined);
+      },
+    };
+  };
+};
+
+function getAppIdAndApiKey(searchClient) {
```

Comment:
maybe add a unit test here. Since we run the tests with algoliasearch v3 and v4 separately, you should be able to cover both branches.

Other case we can add a dev dependency for `"algoliasearch-v3": "npm:algoliasearch@3"` and `import algoliasearch from 'algoliasearc-v3'`

Reply:
TIL: "npm:algoliasearch@3" syntax. Nice! https://github.com/algolia/instantsearch.js/pull/4446/commits/7f49f25ed23470295db24d7c6b0e6d06151ad4e2